### PR TITLE
spirv-fuzz: Fuzzer pass to merge blocks

### PR DESCRIPTION
**Pull request by afd**
This change adds a fuzzer pass that randomly merges blocks in a module.

### CODE REVIEW

File: source/fuzz/fuzzer_pass_merge_blocks.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+void FuzzerPassMergeBlocks::Apply() {
+  // First we populate a sequence of transformations that we might consider
+  // applying.
+  std::vector<TransformationMergeBlocks> potential_transformations;
+  // We do this by considering every block of every function.
+  for (auto& function : *GetIRContext()->module()) {
+    for (auto& block : function) {
+      // We probabilistically decide to ignore some blocks.
+      if (!GetFuzzerContext()->ChoosePercentage(
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_merge_blocks.cpp]: I wonder whether this should be done *after* checking if each potential transformation is applicable.

File: source/fuzz/transformation_merge_blocks.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    opt::IRContext* context, spvtools::fuzz::FactManager* /*unused*/) const {
+  auto second_block = fuzzerutil::MaybeFindBlock(context, message_.block_id());
+  auto first_block =
+      context->cfg()->block(context->cfg()->preds(second_block->id()).at(0));
+
+  auto function = first_block->GetParent();
+  // We need an iterator pointing to the predecessor, hence the loop.
+  for (auto bi = function->begin(); bi != function->end(); ++bi) {
+    if (bi->id() == first_block->id()) {
+      opt::blockmergeutil::MergeWithSuccessor(context, function, bi);
```
Comment (paulthomson) [inline source/fuzz/transformation_merge_blocks.cpp]: Maybe add a comment noting that `CanMergeWithSuccessor` was already called in `IsApplicable`, so we know the precondition holds.

File: source/fuzz/fuzzer_pass_merge_blocks.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_merge_blocks.cpp]: I think the probability of choosing an applicable transformation is the same either way, and it seems preferable to only do the work of checking applicability if we actually want to do take the transformation.

File: source/fuzz/transformation_merge_blocks.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/transformation_merge_blocks.cpp]: Done - via an assertion actually.

**Pull request merged**